mne/minimum_norm/inverse.py

```diff
@@ -1227,6 +1204,87 @@ def apply_inverse_epochs(epochs, inverse_operator, lambda2, method="dSPM",
     return stcs
 
 
+@verbose
+def apply_inverse_cov(cov, info, inverse_operator, nave=1, lambda2=1 / 9,
+                      method="dSPM", pick_ori=None, prepared=False,
+                      label=None, method_params=None, use_cps=True,
+                      verbose=None):
+    """Apply inverse operator to covariance data.
+
+    Parameters
+    ----------
+    cov : Covariance object
+        Covariance data, computed on the time segment for which to compute
+        source power.
+    info : dict
+        The measurement info to specify the channels to include.
+    inverse_operator : instance of InverseOperator
+        Inverse operator.
+    nave : int
+        Number of averages used to regularize the solution.
+    lambda2 : float
+        The regularization parameter.
+    method : "MNE" | "dSPM" | "sLORETA" | "eLORETA"
+        Use minimum norm, dSPM (default), sLORETA, or eLORETA.
+    %(pick_ori)s
+    prepared : bool
+        If True, do not call :func:`prepare_inverse_operator`.
+    label : Label | None
+        Restricts the source estimates to a given label. If None,
+        source estimates will be computed for the entire source space.
+    method_params : dict | None
+        Additional options for eLORETA. See Notes for details.
+    %(use_cps)s
+    %(verbose)s
+
+    Returns
+    -------
+    stc : SourceEstimate | VectorSourceEstimate | VolSourceEstimate
+        The source estimates.
+
+    See Also
+    --------
+    apply_inverse : Apply inverse operator to evoked object.
+    apply_inverse_raw : Apply inverse operator to raw object.
+    apply_inverse_epochs : Apply inverse operator to epochs object.
+
+    Notes
+    -----
+    .. versionadded:: 0.20
+    """
+    _validate_type(cov, Covariance, cov)
+    _validate_type(inverse_operator, InverseOperator, 'inverse_operator')
+    sel = _pick_channels_inverse_operator(cov['names'], inverse_operator)
+    use_names = [cov['names'][idx] for idx in sel]
+    info = pick_info(
+        info, pick_channels(info['ch_names'], use_names, ordered=True))
+    evoked = EvokedArray(
+        np.eye(len(info['ch_names'])), info, nave=nave, comment='cov')
+    is_free_ori = (inverse_operator['source_ori'] == FIFF.FIFFV_MNE_FREE_ORI)
+    if is_free_ori and pick_ori is None:
+        use_ori = 'vector'
+        combine = True
+    else:
+        use_ori = pick_ori
+        combine = False
+    stc = apply_inverse(
+        evoked, inverse_operator, lambda2, method, use_ori, prepared, label,
+        method_params, use_cps=use_cps)
+    # apply (potentially rotated in the vector case) operator twice
+    K = np.reshape(stc.data, (-1, stc.data.shape[-1]))
+    sol = cov.data[sel][:, sel] @ K.T
+    sol = np.sum(K * sol.T, axis=1, keepdims=True)
```

Comment:
I feel like this is not equivalent to
`np.diag(K @ cov.data[sel][:, sel] @ K.T)`
Which i think is what we want. Doesn't the current code include off-diagonal elements for the src covariance matrix?

Reply:
> I feel like this is not equivalent to ...

It should be equivalent, though admittedly it is confusing because it looks like it might be computing something different. The diagonal entries of `A @ B` are given by `(A * B.T).sum(axis=1)`, i.e., the sum of the first-row-of-A-times-first-col-of-B, sum of second-row-of-A-times-second-col-of-B, etc. Just to check it numerically, this code works fine locally if I add it:
```
    sol = np.sum(K * sol.T, axis=1, keepdims=True)
    sol2 = np.diag(K @ cov.data[sel][:, sel] @ K.T)[:, np.newaxis]
    sol3 = np.sum((K @ cov.data[sel][:, sel]) * K, axis=1, keepdims=True)
    np.testing.assert_allclose(sol, sol2)
    np.testing.assert_allclose(sol, sol3)
```

> Doesn't the current code include off-diagonal elements for the src covariance matrix?

Nope, just the diagonal. The end result is has shape `(n_src, 1)`.